Piped/streamed uploads/overwrites fail for zero-length streams

E.g. ``echo -n | acdcli stream emptyfile /`` will fail with
```
15-09-04 22:22:18.364 [ERROR] [acd_cli] - Error uploading stream. Code: 400, msg: {"message":"Bytes not received for request."}
```

while  ``echo | acdcli stream emptyfile /`` will work.

--- a/acdcli/api/content.py
+++ b/acdcli/api/content.py
@@ -81,6 +81,13 @@ def _multipart_stream(metadata: dict, stream, boundary: str, read_callbacks=None
     yield str.encode('\r\n--%s--\r\n' % boundary +
                      'multipart/form-data; boundary=%s' % boundary)
 
+def _stream_is_empty(stream) -> bool:
+    try:
+        return not stream.peek(1)
+    except AttributeError:
+        logger.debug('Stream does not support peeking, upload will fail if stream does '
+                     'not contain at least one byte.')
+        return False
 
 class ContentMixin(object):
     """Implements content portion of the ACD API."""
@@ -171,6 +178,9 @@ class ContentMixin(object):
         """:param stream: readable object
         :param parent: parent node id, defaults to root node if None"""
 
+        if _stream_is_empty(stream):
+            return self.create_file(file_name, parent)
+
         params = {} if deduplication else {'suppress': 'deduplication'}
 
         metadata = {'kind': 'FILE', 'name': file_name}
@@ -214,6 +224,9 @@ class ContentMixin(object):
         """Overwrite content of node with ID *node_id* with content of *stream*.
 
         :param stream: readable object"""
+
+        if _stream_is_empty(stream):
+            return self.clear_file(node_id)
 
         metadata = {}
         import uuid

--- a/tests/test_api_live.py
+++ b/tests/test_api_live.py
@@ -104,6 +104,16 @@ class APILiveTestCase(unittest.TestCase):
 
         self.acd_client.move_to_trash(n['id'])
 
+    def test_upload_stream_empty(self):
+        empty_stream = io.BufferedReader(io.BytesIO())
+        fn = gen_rand_nm()
+
+        n = self.acd_client.upload_stream(empty_stream, fn, parent=None)
+        self.assertEqual(n['contentProperties']['md5'], 'd41d8cd98f00b204e9800998ecf8427e')
+        self.assertEqual(n['contentProperties']['size'], 0)
+
+        self.acd_client.move_to_trash(n['id'])
+
     def test_overwrite(self):
         f, sz = gen_temp_file()
         h = hashing.IncrementalHasher()
@@ -128,6 +138,11 @@ class APILiveTestCase(unittest.TestCase):
         n = self.acd_client.overwrite_stream(s, n['id'], [h.update])
         self.assertEqual(n['contentProperties']['md5'], h.get_result())
         self.assertEqual(n['contentProperties']['size'], sz)
+
+        empty_stream = io.BufferedReader(io.BytesIO())
+        n = self.acd_client.overwrite_stream(empty_stream, n['id'])
+        self.assertEqual(n['contentProperties']['md5'], 'd41d8cd98f00b204e9800998ecf8427e')
+        self.assertEqual(n['contentProperties']['size'], 0)
 
         self.acd_client.move_to_trash(n['id'])
 

